### PR TITLE
fix: repair deprecation hygiene found by the stability-policy audit

### PR DESCRIPTION
**Pull request by ducha-aiki**
Relates to #3895 (the stability policy this audit tested against).

Audited the historical deprecation pattern against the drafted policy in #3896. **Verdict: the pattern was broken in 0.8.3, and there were three hygiene defects.** This PR fixes what is fixable without product decisions:

**1. The 0.8.3 `kornia.utils` removal broke 20 of 34 public names with zero deprecation window.** v0.8.2's `kornia.utils.__all__` had 34 names; 0.8.3 (PR #3449) kept warning shims for only 14 and hard-deleted all deep submodule paths (`kornia.utils.image`, `.draw`, `.grid`, ...). This PR restores warning shims for the 10 names whose implementations still exist elsewhere: `eye_like`, `vec_like`, `safe_solve_with_mask`, `safe_inverse_with_mask`, `is_mps_tensor_safe`, `dataclass_to_dict`, `dict_to_dataclass`, `image_list_to_tensor`, `ImageToTensor`, `CachedDownloader` (lazy, to keep onnx out of the import path) — with tests in `tests/utils/test_deprecated.py`.

**Not restorable (deleted from the library entirely, maintainer decision needed):** `batched_forward`, `torch_meshgrid`, `get_sample_images`, `map_location_to_cpu`. Either resurrect them or list them as intentional hard removals in the 0.9 release notes. Notably **the docs landing page still called `kornia.utils.sample.get_sample_images()`** — broken example since January; fixed here with a plain tensor.

**2. Garbled deprecation version `6.9.10`.** The filters `_t` aliases (`gaussian_blur2d_t`, `get_gaussian_kernel{1,2,3}d_t`) have told users "Since kornia 6.9.10" since Feb 2023 (#2187). Corrected to `0.6.10`. (These have now warned for 2+ minor releases and are removal-eligible under the policy whenever you choose.)

**3. Four-year-old transition warning in `transform_bbox`.** Since Dec 2021 (#1414) every default-argument call on xyxy/xywh boxes warned "will be removed in future versions... set `restore_coordinates=True` as an acknowledgement" — while `None` already behaved exactly like `True`. The transition completed years ago; the warning is pure noise and is removed (zero behavior change; `restore_coordinates=False` still available and documented).

**Also checked, no action needed:** `angle_axis_*` aliases (deprecated 0.7.0 — compliant, removal-eligible); runtime warns in homography/feature/augmentation modules are all conditional on genuine anomalies (SVD non-convergence, empty LAFs, geometric ops in PatchSequential) — justified, kept.

**Verification:**
```
$ pixi run -e default uv run pytest tests/utils/test_deprecated.py tests/test_import.py -q
22 passed
$ pixi run -e default uv run pytest tests/geometry/test_bbox.py -q --dtype=float32
13 passed
$ pixi run lint -> passed
```

Posted on behalf of @ducha-aiki by Claude (Fable 5).

### CODE REVIEW

**Comment by ducha-aiki** [comment]
Second commit (32775db6) completes the repair of the 0.8.3 hard-deletions, per maintainer request:

**Restored with real new homes** (public API again, docs .rst entries added, `kornia.utils` shims warn and forward):
- `get_sample_images` → `kornia.io.get_sample_images` (ported from v0.8.2; the docs landing page uses it again). Verified end-to-end: downloads and loads panda.jpg → `(3, 510, 1020)` float32.
- `batched_forward` → `kornia.core.utils.batched_forward` (doctest included and passing).

**Restored as deprecated-only shims** (obsolete; graceful exit instead of AttributeError):
- `torch_meshgrid` → points at `torch.meshgrid(..., indexing=...)` (the wrapper predates torch 1.10).
- `map_location_to_cpu` → points at `torch.load(..., map_location="cpu")`.

**Systemic guard so this class of break cannot recur:** `tests/test_api_surface.py` + `tests/api_surface.json` — a checked-in inventory of public names for the 12 stable-core modules (712 names). Removing any public symbol now fails CI with a message citing the stability policy, until the inventory is deliberately edited in the same PR. Additions never fail; regenerate with `python tests/test_api_surface.py`.

```
$ pixi run -e default uv run pytest tests/test_api_surface.py tests/utils/test_deprecated.py -q
37 passed
$ pixi run -e default uv run pytest --doctest-modules kornia/core/utils.py -q
1 passed
$ pixi run lint → passed
$ python -c "import warnings, torch; warnings.simplefilter('error'); import kornia" → clean
```

*— comment written by Claude (Claude Code) on behalf of @ducha-aiki*

File: kornia/utils/__init__.py [hunk 1]
```diff
@@ -24,6 +24,27 @@
 from typing import Any
 
 from kornia.core._compat import deprecated
+from kornia.core.ops import (
+    eye_like as _eye_like,
+)
+from kornia.core.ops import (
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/utils/__init__.py:32]: Fixed in 1f8fcb94 — from __future__ import annotations added.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: tests/utils/test_deprecated.py [hunk 2]
```diff
@@ -115,3 +115,89 @@ def test_multiple_calls_each_warn(self):
             utils.create_meshgrid(2, 2)
         dep_warnings = [w for w in caught if issubclass(w.category, DeprecationWarning)]
         assert len(dep_warnings) == 2
+
+
+class TestRestoredWrappers:
+    """Names removed in 0.8.3 without a deprecation window, restored as warning shims.
+
+    See https://github.com/kornia/kornia/pull/3891 follow-up audit: of the 34 public
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply tests/utils/test_deprecated.py:123]: Pushing back: BaseTester is for device/dtype-parametrized tensor-op tests (smoke/cardinality/gradcheck cycles). These are warning-emission and API-surface tests with no device or dtype axis; the recently merged TestDeprecatedWrappers in the same file uses the same plain-class style, and consistency within the file wins.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: docs/source/index.rst [hunk 3]
```diff
@@ -19,15 +19,15 @@ DexiNed edge detection model.
 
 .. code-block:: python
 
-      image = kornia.utils.sample.get_sample_images()[0][None]
+      image = torch.rand(1, 3, 320, 320)
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply docs/source/index.rst]: Moot as described (the torch.rand placeholders were replaced by the restored kornia.io.get_sample_images in 32775db6), but the underlying point was real and went further: the snippets also called DexiNedBuilder, which has never existed in the codebase. Fixed in 1f8fcb94 — all three landing-page snippets now import what they use and call real APIs (kornia.contrib EdgeDetectorBuilder / RTDETRDetectorBuilder / BoxMotTracker).

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: kornia/utils/__init__.py [hunk 4]
```diff
@@ -194,18 +221,139 @@ def one_hot(*args: Any, **kwargs: Any) -> Any:
...
+    """Deprecated: Use `kornia.image.image_list_to_tensor` instead."""
+    return _image_list_to_tensor(*args, **kwargs)
+
+
+@deprecated(
+    replace_with="kornia.image.ImageToTensor",
+    version="0.8.3",
+    extra_reason=" The `kornia.utils` module has been removed. Import from `kornia.image` instead.",
+)
+class ImageToTensor(_ImageToTensor):
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/utils/__init__.py]: Confirmed and fixed in 1f8fcb94 — verified that @deprecated returns a plain class_wrapper function for classes (kornia/core/_compat.py:213), which would have broken isinstance and subclassing. ImageToTensor is now a real subclass that emits the warning from __init__ via _emit_deprecation_warning; added a test asserting it stays a type, is a subclass of kornia.image.ImageToTensor, and instances pass isinstance.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: kornia/utils/__init__.py [hunk 5]
```diff
@@ -194,18 +221,139 @@ def one_hot(*args: Any, **kwargs: Any) -> Any:
...
+    """Deprecated: Use `kornia.image.ImageToTensor` instead."""
+
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/utils/__init__.py:387]: Fixed in 1f8fcb94 exactly as suggested — real subclass, warning in __init__, with a regression test for type identity.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: docs/source/index.rst [hunk 6]
```diff
@@ -19,15 +19,15 @@ DexiNed edge detection model.
...
       model = DexiNedBuilder.build()
       model.save(image)
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply docs/source/index.rst]: Fixed in 1f8fcb94 — DexiNedBuilder (nonexistent) replaced with kornia.contrib.EdgeDetectorBuilder, and all three snippets now carry explicit imports. Also note the snippets use the restored kornia.io.get_sample_images rather than torch.rand.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: docs/source/index.rst [hunk 7]
```diff
@@ -19,15 +19,15 @@ DexiNed edge detection model.
...
 
 RTDETRDetector for object detection.
 
 .. code-block:: python
 
-      image = kornia.utils.sample.get_sample_images()[0][None]
+      image = torch.rand(1, 3, 320, 320)
       model = RTDETRDetectorBuilder.build()
       model.save(image)
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply docs/source/index.rst]: Fixed in 1f8fcb94 — explicit imports added to all three landing-page snippets.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: kornia/utils/__init__.py [hunk 8]
```diff
@@ -194,18 +221,139 @@ def one_hot(*args: Any, **kwargs: Any) -> Any:
...
+        import warnings
+
+        from kornia.onnx.download import CachedDownloader as _CachedDownloader
+
+        warnings.warn(
+            "`kornia.utils.CachedDownloader` is deprecated since kornia 0.8.3."
+            " Use `kornia.onnx.download.CachedDownloader` instead.",
+            DeprecationWarning,
+            stacklevel=2,
+        )
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/utils/__init__.py]: Fixed in 1f8fcb94 — the lazy CachedDownloader shim now goes through _emit_deprecation_warning, matching the always-warn filter behavior of every other shim.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: docs/source/core.rst [hunk 9]
```diff
@@ -8,5 +8,7 @@ kornia.core
 .. currentmodule:: kornia.core
 
 .. autoclass:: TensorWrapper
+
+.. autofunction:: kornia.core.utils.batched_forward
     :members:
     :undoc-members:
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply docs/source/core.rst:12]: Confirmed and fixed in the latest commit — my insertion had severed the options from the autoclass directive. :members:/:undoc-members: are back under TensorWrapper, autofunction follows after; docs build verified clean.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

File: kornia/core/utils.py [hunk 10]
```diff
@@ -397,3 +397,51 @@ def dict_to_dataclass(dict_obj: Dict[str, Any], dataclass_type: Type[T]) -> T:
...
+        n_batches = int(B // bs + 1)
+        for batch_idx in range(n_batches):
+            st = batch_idx * bs
+            end = min((batch_idx + 1) * bs, B)
+            if st >= end:
+                continue
+            out_list.append(model_dev(data[st:end].to(device), **kwargs))
+        out = torch.cat(out_list, 0)
+        return out.to(data.device)
+    return model(data, **kwargs)
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/core/utils.py]: Confirmed and fixed — the small-batch branch now runs the same transfers as the micro-batch branch (`model_dev(data.to(device), **kwargs).to(data.device)`). This bug existed verbatim in the v0.8.2 original; verified the fixed contract with a parameterized Linear model.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*

**Comment by ducha-aiki** [comment]
The Windows failure is not from this PR's changes — it's the strict import gate merged in #3894 meeting GitHub's windows-2025 runner rollout: `import kornia` eagerly imports onnxruntime (via `kornia.feature.lightglue_onnx`), and onnxruntime warns at import on Windows Server 2025, which the gate promotes to an error. #3894's own CI passed only because its runners still drew Server 2022; main is currently red on any windows-2025 runner.

Hotfix: #3899 (pre-imports onnxruntime alongside torch before arming the error filter, per the gate's own third-party scoping). Merge #3899 first, then re-run this PR's failed Windows jobs.

*— comment written by Claude (Claude Code) on behalf of @ducha-aiki*

File: kornia/utils/__init__.py [hunk 11]
```diff
@@ -194,18 +223,195 @@ def one_hot(*args: Any, **kwargs: Any) -> Any:
...
+            "kornia.onnx.download.CachedDownloader",
+            "0.8.3",
+            " The `kornia.utils` module has been removed. Import from `kornia.onnx.download` instead.",
+        )
+        return _CachedDownloader
+    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
+
+
 __all__ = [
+    "CachedDownloader",
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/utils/__init__.py:455]: Confirmed for 5 of the 6 and fixed in fb07eb2c: warning shims + tests added for `get_cuda_device_if_available`, `get_mps_device_if_available`, `get_cuda_or_mps_device_if_available`, `is_autocast_enabled`, and `xla_is_available`. The sixth, `deprecated`, was never broken — it is imported at the top of `kornia/utils/__init__.py` and has resolved as `kornia.utils.deprecated` all along (it was only absent from `__all__`).

To close the audit rather than iterate it: the full v0.8.2 `__all__` (34 names, verified against `git show v0.8.2:kornia/utils/__init__.py`) is now frozen as a completeness test (`test_v082_public_surface_fully_restored`) asserting every public name still resolves — only the underscore-private `_extract_device_dtype` is excluded by policy.

*— reply written by Claude (Claude Code) on behalf of @ducha-aiki*
